src/Altinn.Platform/Altinn.Platform.Storage/Storage/Helpers/AuthorizeInstancesHelper.cs

```diff
@@ -0,0 +1,289 @@
+            XacmlJsonResponse response = await _pdp.GetDecisionForRequest(xacmlJsonRequest);
+
+            foreach (XacmlJsonResult result in response.Response)
+            {
+                if (DecisionHelper.ValidateDecisionResult(result, user))
+                {
+                    string instanceId = string.Empty;
+                    string actiontype = string.Empty;
+
+                    // Loop through all attributes in Category from the response
+                    foreach (XacmlJsonCategory category in result.Category)
+                    {
+                        var attributes = category.Attribute;
+
+                        foreach (var attribute in attributes)
+                        {
+                            if (attribute.AttributeId.Equals(XacmlResourceActionId))
+                            {
+                                actiontype = attribute.Value;
+                            }
+
+                            if (attribute.AttributeId.Equals(AltinnXacmlUrns.InstanceId))
+                            {
+                                instanceId = attribute.Value;
+                            }
+                        }
+                    }
+
+                    // Find the instance that has been validated to add it to the list of authorized instances.
+                    Instance authorizedInstance = instances.FirstOrDefault(i => i.Id == instanceId);
+
+                    // Checks if it has already been authorized.
+                    int instancePosition = InstanceAlreadyAuthorized(authorizedInstances, authorizedInstance);
+
+                    // The instance has already been added to the authorizedInstance list if the position is 0 or above. 
+                    if (instancePosition >= 0)
+                    {
+                        // Only need to check if the action type is write, because read do not add any special rights to the MessageBoxInstane.
+                        if (actiontype.Equals("write"))
+                        {
+                            MessageBoxInstance authorizedMsgBoxInstance = authorizedInstances[instancePosition];
+                            authorizedMsgBoxInstance.AuthorizedForWrite = true;
+                            authorizedMsgBoxInstance.AllowDelete = true;
+                        }
+                    }
+                    else
+                    {
+                        MessageBoxInstance messageBoxInstance = InstanceHelper.ConvertToMessageBoxInstance(authorizedInstance);
+
+                        if (actiontype.Equals("write"))
+                        {
+                            messageBoxInstance.AuthorizedForWrite = true;
+                            messageBoxInstance.AllowDelete = true;
+                        }
+
+                        authorizedInstances.Add(messageBoxInstance);
+                    }
+                }
+            }
```

Comment:
this section could probably be simplified by linq. That way you do not need to know the position of the element.